File: uhrwerk-backend-mysql/src/test/java/io/qimia/uhrwerk/dao/ConnectionDAOTest.java
package io.qimia.uhrwerk.dao;

import io.qimia.uhrwerk.ConnectionHelper;
import io.qimia.uhrwerk.common.metastore.config.ConnectionResult;
import io.qimia.uhrwerk.common.metastore.config.ConnectionService;
import io.qimia.uhrwerk.common.model.Connection;
import io.qimia.uhrwerk.common.model.ConnectionType;
import io.qimia.uhrwerk.config.ConnectionBuilder;
import org.apache.log4j.Logger;
import org.junit.jupiter.api.Test;

import java.sql.PreparedStatement;
import java.sql.SQLException;
import java.util.ArrayList;

import static org.junit.jupiter.api.Assertions.*;

class ConnectionDAOTest {

  java.sql.Connection db;
  ConnectionService service;
  String[] connNames = {"S3", "JDBC", "file"};
  private Logger logger;

  @org.junit.jupiter.api.BeforeEach
  void setUp() throws SQLException {
    db =
            ConnectionHelper.getConnection();
    service = new ConnectionDAO(db);
    logger = Logger.getLogger(this.getClass());
  }

  @org.junit.jupiter.api.AfterEach
  void tearDown() throws SQLException {
    if (db != null) if (!db.isClosed()) db.close();
  }

  @Test
  void insert() {
    Connection conn = new Connection();
    conn.setName("Test-Conn1");
    conn.setType(ConnectionType.FS);
    conn.setPath("/some/path/test1");
    conn.setKey();
    ConnectionResult result = service.save(conn, true);
    assertTrue(result.isSuccess());
    assertNotNull(result.getNewConnection());
    assertNotNull(result.getNewConnection().getId());
    logger.info(result.getNewConnection());
  }

  @Test
  void update() {
    Connection conn = new Connection();
    conn.setName("Test-Conn1");
    conn.setType(ConnectionType.S3);
    conn.setPath("/some/path/updated");
    conn.setAwsAccessKeyID("access-key-id1");
    conn.setAwsSecretAccessKey("secret-access-key1");
    conn.setKey();
    ConnectionResult result = service.save(conn, true);
    assertTrue(result.isSuccess());
    assertNotNull(result.getNewConnection());
    assertNotNull(result.getNewConnection().getId());
    logger.info(result.getNewConnection());
  }

  @Test
  void insertFullConnection() {

    var conns = new ArrayList<Connection>();
    var connection1 =
        (new ConnectionBuilder())
            .name(connNames[0])
            .s3()
            .path("S3Path")
            .secretId("ID")
            .secretKey("key")
            .done()
            .build();
    conns.add(connection1);

    var connection2 =
        (new ConnectionBuilder())
            .name(connNames[1])
            .jdbc()
            .jdbcUrl("url")
            .jdbcDriver("driver")
            .user("user")
            .pass("pass2")
            .done()
            .build();
    conns.add(connection2);

    var connection3 =
        (new ConnectionBuilder()).name(connNames[2]).file().path("filePath").done().build();
    conns.add(connection3);

    for (Connection c : conns) {
      ConnectionResult result = service.save(c, true);
      assertTrue(result.isSuccess());
      assertNotNull(result.getNewConnection());
      assertNotNull(result.getNewConnection().getId());
      logger.info(c);
      logger.info(result.getNewConnection());
    }
  }

  @Test
  void getByNameConnection() {
    try {
      for (String n : connNames) {
        Connection con = service.getByName(n);
        logger.info(con);
        assertEquals(n, con.getName());
      }
    } catch (SQLException e) {
      logger.info("caught in Connection.");
    }
  }

  @Test
  void getByIDConnection() {
    try {
      for (String n : connNames) {
        Connection testCon = new Connection();
        testCon.setName(n);
        testCon.setKey();
        Connection con = service.getById(testCon.getId());
        logger.info(con);
        assertEquals(testCon.getId(), con.getId());
      }
    } catch (SQLException e) {
      logger.info("caught in Connection.");
    }
  }

  @Test
  void getDependency() {}

  @Test
  void getByStatementConnection() {
    try {
      PreparedStatement statement =
          db.prepareStatement("SELECT * FROM CONNECTION WHERE NAME in ();");
      Connection[] con = service.getConnections(statement);
      for (int i = 0; i < con.length; i++) {
        logger.info(con[i]);
        assertEquals(connNames[i], con[i].getName());
      }
    } catch (SQLException e) {
      logger.info("caught in Connection.");
    }
  }
}

File: uhrwerk-backend-mysql/src/test/kotlin/io/qimia/uhrwerk/dao/ConnectionDAOTest.kt
package io.qimia.uhrwerk.dao

import io.qimia.uhrwerk.TestHelper
import io.qimia.uhrwerk.common.model.ConnectionModel
import io.qimia.uhrwerk.config.ConnectionBuilder
import io.qimia.uhrwerk.repo.HikariCPDataSource
import io.qimia.uhrwerk.TestData
import org.junit.jupiter.api.*
import org.slf4j.LoggerFactory
import org.testcontainers.containers.MySQLContainer
import org.testcontainers.containers.output.Slf4jLogConsumer
import org.testcontainers.junit.jupiter.Container
import org.testcontainers.junit.jupiter.Testcontainers

@Testcontainers
internal class ConnectionDAOTest {

    private val service = ConnectionDAO()

    @AfterEach
    fun cleanUp() {
        TestHelper.cleanData("CONNECTION", LOGGER)
    }

    @BeforeEach
    fun saveData() {
    }

    @Test
    fun save() {
        val result = service.save(TestData.connection("Connection-Test"), true)
        Assertions.assertTrue(result.isSuccess)
        Assertions.assertNotNull(result.newConnection)
        Assertions.assertNotNull(result.newConnection.id)
        LOGGER.info(result.newConnection!!.toString())
    }


    @Test
    fun insertFullConnection() {
        val conns = mutableListOf<ConnectionModel>()

        val connection1 = ConnectionBuilder()
            .name(CONN_NAMES[0])
            .s3()
            .path("S3Path")
            .secretId("ID")
            .secretKey("key")
            .done()
            .build()
        conns.add(connection1)

        val connection2 = ConnectionBuilder()
            .name(CONN_NAMES[1])
            .jdbc()
            .jdbcUrl("url")
            .jdbcDriver("driver")
            .user("user")
            .pass("pass2")
            .done()
            .build()

        conns.add(connection2)

        val connection3 =
            ConnectionBuilder().name(CONN_NAMES[2]).file().path("filePath").done().build()
        conns.add(connection3)

        conns.forEach {
            LOGGER.info(it.toString())
            val result = service.save(it, true)
            Assertions.assertTrue(result.isSuccess)
            Assertions.assertNotNull(result.newConnection)
            Assertions.assertNotNull(result.newConnection.id)
            LOGGER.info(result.newConnection.toString())
        }

    }

    companion object {
        private val CONN_NAMES = arrayOf("S3", "JDBC", "file")

        private val LOGGER = LoggerFactory.getLogger(ConnectionDAOTest::class.java)

        @Container
        var MY_SQL_DB: MySQLContainer<*> = TestHelper.mysqlContainer()

        @BeforeAll
        @JvmStatic
        fun setUp() {
            val logConsumer = Slf4jLogConsumer(LOGGER)
            MY_SQL_DB.followOutput(logConsumer)
            HikariCPDataSource.initConfig(
                MY_SQL_DB.jdbcUrl,
                MY_SQL_DB.username,
                MY_SQL_DB.password
            )
        }

        @AfterAll
        @JvmStatic
        fun tearDown() {
            HikariCPDataSource.close()
        }
    }
}